### ISSUE TITLE
[Bug]:  Savestate problem

### ISSUE THREAD

**Issue by TheAftermathDestroyer** [visualboyadvance-m/visualboyadvance-m#1184]
### Expected behavior?

When I try loading state on a game, it should fully load state

### Actual behavior?

When I try loading state, something off happens. I try loading state then the game breaks. I'm switching to mGBA >:(

### Steps to reproduce the problem

1. I load a game
2. Play around with the game and save state before running into a problem
3. I run into a problem
4.  I try to load state
5. The game gets broken
...

### Error messages?

There were no error messages. The save state loaded but with some bugs.

### What game were you playing

<Some of this information can be provided with a screenshot of Rom Information, SAVE TYPE for GBA titles can be obtained from Options > Game Boy Advance > Config>
Title: Super Mario Land 2
GB, GBC or GBA: GB
ROM CODE: I don't know
BIOS Used?: GameBoy
SAVE TYPE: [Super Mario Land 2 DX (Hack v1.8.1)-01.sgm.zip](https://github.com/visualboyadvance-m/visualboyadvance-m/files/12752515/Super.Mario.Land.2.DX.Hack.v1.8.1.-01.sgm.zip)
Clean, Intro'd or Hacked?:
[SML2DXv181.zip](https://github.com/visualboyadvance-m/visualboyadvance-m/files/12752523/SML2DXv181.zip)

### What settings are you using?

Video driver (simple, opengl): OpenGL
Sound driver (directsound, xaudio, openal, SDL): SDL
Scale Filter Used?: None

### Specifications and Build details

VBA-M: (include commit version number and whether you are using the 32-bit or 64-bit version) 64 bit
OS: (include version and OS build number using WinVer if on Windows): Microsoft Windows 10 [Version 10.0.19045.3448]
If using Linux, specify if using xorg or Wayland:
CPU: Intel Core i5
GPU: Intel® Core™ i5 i5-3230M 2.6 GHz
RAM: 8GB

### Copyright Compliance

- [X] There are no infringing files attached to this bug report.

**Comment by ZachBacon** [comment]
Fill out the details or I'm closing this.

**Comment by TheAftermathDestroyer** [comment]
> Fill out the details or I'm closing this.

Check again.

**Comment by Squall-Leonhart** [comment]
the GBC core doesn't load states made from older versions of the emulator correctly in some cases, either producing audio issues or the colours are completely off.

**Comment by Squall-Leonhart** [comment]
and the DX hack for this game is not running in GB mode, it runs in GBC.

**Comment by TheAftermathDestroyer** [comment]
Thanks! The savestates are working now! I'd recommend this for anyone having the same problem on VBA-Next.

**Comment by Megasonic999** [comment]
I'm using the original game, and when I load state after losing a powerup, it goes back to where I saved it, but with the player status before loading states. Meanig I still lose a powerup or lose a life.

**Comment by Squall-Leonhart** [comment]
Literally impossible.

**Comment by Megasonic999** [comment]
I know, but still, it happens

**Comment by Megasonic999** [comment]
Rectification, savestates are broken for this game when loading

**Comment by Squall-Leonhart** [comment]
This game is abusing SRAM in some way, in that when we allow the state to change the battery the data is restored correctly. (The data is always saved correctly). If we have the option set to not change the battery data on restore, several counters fail to restore.

We are not the only emulator affected in this way, but we'll need to identify a way to correctly support states in games that behave like this while being able to keep the

**Comment by Squall-Leonhart** [comment]
These games are using part of the eeprom memory as game RAM, and basically not using the GB's own memory at all, so upon loading the state with the do not restore battery data option set you're getting the CPU state loaded and referencing whatever the cart's boot state has put there.

**Comment by endrift** [comment]
GB games do not have EEPROM. Some GB games use part of SRAM as working RAM due to how little RAM the GB has. People have probably been running into this for years.

**Comment by Squall-Leonhart** [comment]
These 2 games are pretty unique in how they don't seem to actively read/write the WRAM and RAM of the gb hardware, they seem to garbage fill these regions with copy pasta'd garbage and execute entirely from SRAM.

Pokemon on the otherhand seems to flip data in and out of SRAM during battles, which might be stat and moveset related, these aren't updating while traversing the overworld.

Best bet is pobably to adopt the same behavior as mgba here and restore the data in memory only, only updating the battery file if an update has been made.

Good point on the eeprom, i'd like to claim typo but it was a slip of the mind.

**Comment by endrift** [comment]
Pokémon uses SRAM as a buffer when decompressing 'mon graphics. It even does this on the title screen.

**Comment by Megasonic999** [comment]
Here ya go man:
SAVE TYPE: 
[Super Mario Land 2 - 6 Golden Coins (USA, Europe)-01.sgm.zip]Super Mario Land 2 - 6 Golden Coins (USA, Europe)-01.sgm.zip(https://github.com/visualboyadvance-m/visualboyadvance-m/files/15213498/Super.Mario.Land.2.-.6.Golden.Coins.USA.Europe.-01.sgm.zip)
Clean, Intro'd or Hacked? 
[Super Mario Land 2 - 6 Golden Coins (USA, Europe).gb.zip]Super Mario Land 2 - 6 Golden Coins (USA, Europe).gb.zip(https://github.com/visualboyadvance-m/visualboyadvance-m/files/15213500/Super.Mario.Land.2.-.6.Golden.Coins.USA.Europe.gb.zip)

**Comment by ZachBacon** [comment]
How many times do I have to say this? DO NOT LINK ROMS TO GITHUB.

**Comment by Megasonic999** [comment]
Sorry. my bad !😅 I saw the link of the rom hack above, so I thought it was okay
Not gonna happen again

**Comment by ZachBacon** [comment]
> Sorry. my bad !😅 I saw the link of the rom hack above, so I thought it was okay
> Not gonna happen again

That was a link to an IPS patch that is applied to a rom. That's ok, but to link to a rom puts the project in jeopardy.

**Comment by Megasonic999** [comment]
Yeah, definitly don't want to repeat the same story with yuzu and it's forks
So anyways, any luck identifying the problem ?
no need to rush though. There are still alternatives to play this game until it's fixed or figured out

**Comment by Squall-Leonhart** [comment]
You can deselect "Do not change battery state" on the Load State menu.

**Comment by Megasonic999** [comment]
Is it in files or options?

**Comment by Megasonic999** [comment]
Never mind, I found it. Thanks.

**Comment by Squall-Leonhart** [comment]
> Pokémon uses SRAM as a buffer when decompressing 'mon graphics. It even does this on the title screen.

Apparently Pokemon GSC uses SRAM for Window data @endrift, loading an in-battle savestate can trigger a white screen crash when  opening and then closing the Pack screen during a fight, and also triggering the"No windows available for popping." debug message can can occur if you load the state without ever having been in a battle prior. [https://github.com/visualboyadvance-m/visualboyadvance-m/issues/1153]

This is likely the cause of https://github.com/visualboyadvance-m/visualboyadvance-m/issues/1123 and is the cause of

**Comment by Squall-Leonhart** [comment]
Depends on: https://github.com/visualboyadvance-m/visualboyadvance-m/issues/781
